core/modules.rs

```diff
@@ -200,150 +201,80 @@ impl std::fmt::Display for ModuleType {
 pub struct ModuleSource {
   pub code: ModuleCode,
   pub module_type: ModuleType,
-  pub module_url_specified: String,
-  pub module_url_found: String,
+  module_url_specified: ModuleName,
+  /// If the module was found somewhere other than the specified address, this will be [`Some`].
+  module_url_found: Option<ModuleName>,
 }
 
-/// Module code can be sourced from strings or bytes that are either owned or borrowed. This enumeration allows us
-/// to perform a minimal amount of cloning and format-shifting of the underlying data.
-///
-/// Note that any [`ModuleCode`] created from a `'static` byte array or string must contain ASCII characters.
-///
-/// Examples of ways to construct a [`ModuleCode`] object:
-///
-/// ```rust
-/// # use deno_core::ModuleCode;
-///
-/// let code: ModuleCode = "a string".into();
-/// let code: ModuleCode = b"a string".into();
-/// ```
-#[derive(Debug)]
-pub enum ModuleCode {
-  /// Created from static data -- must be 100% 7-bit ASCII!
-  Static(&'static [u8]),
-
-  /// An owned chunk of data.
-  Owned(Vec<u8>),
-
-  /// Scripts loaded from the `deno_graph` infrastructure.
-  Arc(Arc<str>),
-}
-
-impl ModuleCode {
-  #[inline(always)]
-  pub fn as_bytes(&self) -> &[u8] {
-    match self {
-      Self::Static(b) => b,
-      Self::Owned(b) => b,
-      Self::Arc(s) => s.as_bytes(),
-    }
-  }
-
-  pub fn try_static_ascii(&self) -> Option<&'static [u8]> {
-    match self {
-      Self::Static(b) => Some(b),
-      _ => None,
-    }
-  }
-
-  /// Takes a [`ModuleCode`] value as an owned [`String`]. May be slow.
-  pub fn take_as_string(self) -> String {
-    match self {
-      Self::Static(b) => String::from_utf8(b.to_vec()).unwrap(),
-      Self::Owned(b) => String::from_utf8(b).unwrap(),
-      Self::Arc(s) => (*s).to_owned(),
+impl ModuleSource {
+  /// Create a [`ModuleSource`] without a redirect.
+  pub fn new(
+    module_type: impl Into<ModuleType>,
+    code: ModuleCode,
+    specifier: ModuleSpecifier,
+  ) -> Self {
+    let module_url_specified = specifier.into();
+    Self {
+      code,
+      module_type: module_type.into(),
+      module_url_specified,
+      module_url_found: None,
     }
   }
 
-  /// Truncates a `ModuleCode`] value, possibly re-allocating or memcpy'ing. May be slow.
-  pub fn truncate(&mut self, index: usize) {
-    match self {
-      Self::Static(b) => *self = Self::Static(&b[..index]),
-      Self::Owned(b) => b.truncate(index),
-      // We can't do much if we have an Arc<str>, so we'll just take ownership of the truncated version
-      Self::Arc(s) => *self = s[..index].to_owned().into(),
+  /// Create a [`ModuleSource`] with a potential redirect.
+  pub fn new_with_maybe_redirect(
```

Comment:
If `specifier_found` is required then it appears to be `new_with_redirect` without the "maybe"

Reply:
In this case the code is doing a compare with the redirect value so I wanted to include the word `maybe` to indicate that it _might_ include a redirect. I can rename the method and improve the comment.